### PR TITLE
Prints validation metric. Closes issue #56.

### PR DESCRIPTION
**Pull request by seba-1511**
This PR adds the `print_metric_score` function to the MLP model, and calls it in the fit method on the validation set, if the `validation_metric` parameter is not None. Any dataset, set, and metric can be used with this method.

While implementing, I was wondering if I should use `set_train_mode(False)`, but as it wasn't used in the `FitPredictError` experiment I didn't.

Also, as I am not comfortable enough with some of the other models (DBNs, Balances, and RBMs) I did not try to add it there. The autoencoder example (which doesn't use the autoencoder class by the way) seemed to work fine. And as I haven't looked carefully at the RNN code yet, I did not try to modify it.

Thanks for your comments and have a nice day,
Séb

### CODE REVIEW

**Comment by ursk** [comment]
HI Séb,
there are two things we do going from training to prediction mode: In predict_generator, there is the `self.data_layer.use_set(setname, predict=True)`, which tells the dataset provider to take batches from the test set. Then in fit_predict_err we have ` self.model.set_train_mode(False)` which switches the drop-out layers, switching them to deterministic mode, and switching batch normalization to inference mode. All of these need to be undone if you want to continue training after running inference. It should be relatively easy to check if you still get the same final error if you have the callback in place, than without it. Also make sure the validation misclass percentage is the same as what you get from the final metrics call, to verify that the dropout layers and batch norm layers switch mode correctly.

**Comment by seba-1511** [comment]
Hello,

Thanks for the comments, I completely forgot to revert back to the train set. I just tested it with mnist small, and with the same random seed I get the same results when both printing and not printing the validation metrics, which should be news.

One thing I believe could still be improved is that after printing the metrics, I revert to the train set. But maybe the user used another set before printing the metric score. From looking at the DataLayer class, I can find a good way to get which set was being used before printing the metric score. Any idea ?

Have a nice day,
Séb

**Comment by ursk** [comment]
Hi Séb,

It great that the feature works with the MNIST-small model; see above though for difficulties I expect with dropout and batch normalization because they rely on `self.model.set_train_mode`. In this case, your feature would lead to unexpected behavior, i.e. you would get wrong metrics reported. Note the MNIST example does not use those layer types.

All dataset have, or should have, a training set, so generally I think it's safe to attempt to revert to that for continued training.

**Comment by seba-1511** [comment]
Alright, for some reason I thought that `set_train_mode(False)` was already in `predict_generator` and removed it in commit 2a6ddd1. I just re-added the line back, and testing with batch norm and dropout gives both times the same results, whether the validation metric is printed or not.

**Comment by scttl** [comment]
I've taken a look through this change and think the following updates are in order:
1.  Revert the changes to `mnist-small.yaml` example.  Probably better to make a new `mnist-epoch_val_print-small.yaml` or something like that.
2.  The current `validation_metric` name is misleading and should be changed.  We already have a `metrics` parameter for things to compute after training, maybe this parameter should instead be called `per_epoch_metrics` and have a similar format to metrics.
3.  From the above we shouldn't be restricted to just specifying a single metric, and only be allowed to compute them on a validation set.  We should support per epoch computation of potentially multiple metrics on each of the train, test, and validation sets.  If this variable is not specified we default to our current training error reporting.
4.  We should add this functionality to RNN's since doing so should be straightforward.

**Comment by seba-1511** [comment]
Good afternoon,

Didn't have much time to work on this. I included the suggested modifications 1, 2, 3 and printing metrics for RNNs is coming soon. (the name of the parameter is now `epoch_metrics`)

As there might be a list of such metrics for each epoch (eg, in mnist-epoch_metrics-small.yaml), I added a "header" stating the epoch for which the metrics were printed.

**Comment by seba-1511** [comment]
Alright, I believe it is also working for RNNs. In addition I modified print_metric_score so that it takes a list of metrics for one set to avoid recomputing the predictions. One thing that could be nifty enhancement would be to print the metrics in a table so that it is easier to see their evolution. I am trying to see if there is a way to do it through python's `logging`.

**Comment by scttl** [comment]
HI Séb,

I've had a look through the updates and they look good from my perspective. :+1:   

As for table based printing, I think the best way to go may be to support dumping these metrics to a user specified flat text file (1 column per metric, one row per epoch).  We have this ability for regular, end of model training metrics via the `-o` parameter and the `dump_metrics()` function.  This computes other info so probably not what we want to use in this case, but may give you a couple of ideas to get started.

Only other suggestion would be to flatten all of your commits down to a single one with relevant description to keep the changelog clean, but I can do that on my side when I merge it in.  Let me know if I should go ahead and merge now, or wait for some table file dumping changes (maybe making these a separate pull request would be better anyway).

Thanks for the contribution!

**Comment by seba-1511** [comment]
Hello,

I did not know about `-o`, that can be useful. If I do the table printout, it will be quite further in the future and so I believe it can wait for its own pull request.

Thank you for the comments and your help,
Séb
